Mathlib/Topology/Constructible.lean

```diff
@@ -0,0 +1,345 @@
+  | union s hs t ht hs' ht' => rw [image_union]; exact hs'.union ht'
+  | compl s hs hs' =>
+    rw [← range_diff_image hf.injective]
+    exact (hfcomp.isConstructible hf.isClosed_range.isOpen_compl).of_compl.sdiff hs'
+
+section CompactSpace
+variable [CompactSpace X] {P : ∀ s : Set X, IsConstructible s → Prop} {B : Set (Set X)}
+
+lemma IsRetroCompact.isCompact (hs : IsRetroCompact s) : IsCompact s := by
+  simpa using hs CompactSpace.isCompact_univ
+
+lemma TopologicalSpace.IsTopologicalBasis.isRetroCompact_iff_isCompact
+    (basis : IsTopologicalBasis B) (compact_inter : ∀ U ∈ B, ∀ V ∈ B, IsCompact (U ∩ V))
+    (hU : IsOpen U) : IsRetroCompact U ↔ IsCompact U := by
+  refine ⟨IsRetroCompact.isCompact, fun hU' {V} hV' hV ↦ ?_⟩
+  -- have hb (i : PLift ι) : IsCompact (b i.down) := by simpa using compact_inter i.down i.down
+  -- have := isCompact_open_iff_eq_finite_iUnion_of_isTopologicalBasis (b ∘ PLift.down)
+  --   (by simpa [PLift.down_surjective.range_comp] using basis) hb
+  obtain ⟨s, rfl⟩ := eq_sUnion_finset_of_isTopologicalBasis_of_isCompact_open _ basis _ hU' hU
+  obtain ⟨t, rfl⟩ := eq_sUnion_finset_of_isTopologicalBasis_of_isCompact_open _ basis _ hV' hV
+  rw [Set.sUnion_inter_sUnion]
+  refine ((s.finite_toSet.image _).prod (t.finite_toSet.image _)).isCompact_biUnion ?_
+  simp only [mem_prod, mem_image, Finset.mem_coe, Subtype.exists, exists_and_right, exists_eq_right,
+    and_imp, forall_exists_index, Prod.forall]
+  exact fun u v hu _ hv _ ↦ compact_inter _ hu _ hv
+
+lemma TopologicalSpace.IsTopologicalBasis.isRetroCompact (basis : IsTopologicalBasis B)
+    (compact_inter : ∀ U ∈ B, ∀ V ∈ B, IsCompact (U ∩ V)) (hU : U ∈ B) : IsRetroCompact U :=
+  (basis.isRetroCompact_iff_isCompact compact_inter <| basis.isOpen hU).2 <| by
+    simpa using compact_inter _ hU _ hU
+
+lemma TopologicalSpace.IsTopologicalBasis.isConstructible (basis : IsTopologicalBasis B)
+    (compact_inter : ∀ U ∈ B, ∀ V ∈ B, IsCompact (U ∩ V)) (hU : U ∈ B) : IsConstructible U :=
+  (basis.isRetroCompact compact_inter hU).isConstructible <| basis.isOpen hU
+
+proof_wanted IsConstructible.induction_of_isTopologicalBasis (basis : IsTopologicalBasis B)
+    (compact_inter : ∀ U ∈ B, ∀ V ∈ B, IsCompact (U ∩ V))
+    (sdiff : ∀ U (hU : U ∈ B) 𝒱 (h𝒱B : 𝒱 ⊆ B) (h𝒱 : 𝒱.Finite), P (U \ ⋃ V ∈ 𝒱, V)
+      ((basis.isConstructible compact_inter hU).sdiff <| .biUnion h𝒱 fun V hV ↦
+        basis.isConstructible compact_inter <| h𝒱B hV))
+    (union : ∀ s hs t ht, P s hs → P t ht → P (s ∪ t) (hs.union ht))
+    (s : Set X) (hs : IsConstructible s) : P s hs
+
+end CompactSpace
+
+/-! ### Locally constructible sets -/
+
+/-- A set in a topological space is locally constructible, if it can be partitioned along an open
+cover such that every part is constructible -/
+@[stacks 005G]
+def IsLocallyConstructible (s : Set X) : Prop :=
+  ∃ 𝒰 : Set (Set X),
+    (∀ x, ∃ U ∈ 𝒰, x ∈ U) ∧ (∀ U ∈ 𝒰, IsOpen U) ∧ ∀ U ∈ 𝒰, IsConstructible (s ∩ U)
```

Comment:
Unless I'm missing something, a subset `s` of an open set `U` is constructible iff it is constructible in `U`. Proof:
* If `s` constructible, say `s = U₁ ∩ C₁ ∪ ... ∪ Uₙ ∩ Cₙ`, then `s = (U₁ ∩ U) ∩ (C₁ ∩ U) ∪ ... ∪ (Uₙ ∩ U) ∩ (Cₙ ∩ U)` is constructible in `U`.
* If `s` constructible in `U`, say `s = (U₁ ∩ U) ∩ (C₁ ∩ U) ∪ ... ∪ (Uₙ ∩ U) ∩ (Cₙ ∩ U)`, then `s = (U₁ ∩ U) ∩ C₁ ∪ ... ∪ (Uₙ ∩ U) ∩ Cₙ` is constructible

Reply:
In fact, I have already proved this as `IsConstructible.image_of_isOpenEmbedding` and around. I guess I need some retrocompactness assumption. Is this what you had in mind?

Reply:
Yeah I think when working under full generality, in your second bullet point the `U₁ ∩ U` and `U \ C₁` needs to be retrocompact, but being retrocompact in `U` may not imply being retrocompact in `X`.
If every open cover has a refinement by retrocompact opens, then I think your definition of IsLocallyConstructible agrees with the usual one. However, it seems a scheme doesn't necessarily satisfy this condition. Were you able to prove Chevalley (image of locally constructible under finite presented scheme morphism is locally constructible) in full generality using your definition?

Reply:
We proved Chevalley in full generality for constructible sets, but not for locally constructible sets, no

Reply:
It really looks like it is Grothendieck's definition of a locally constructible set, see EGA, 0 III, 9.1.11.
It might not be if notation has a distinct meaning than the one I imagine.

http://www.numdam.org/item/PMIHES_1961__11__5_0.pdf

Reply:
`s ↓∩ U` is a set in `s`, not in `U`.

Reply:
🤦